security/s2a/internal/crypter/counter_test.go

```diff
@@ -0,0 +1,105 @@
+package crypter
+
+import (
+	"math"
+	"testing"
+)
+
+// counterFromValue creates a new counter given an initial value.
+func counterFromValue(value uint64) counter {
+	c := newCounter()
+	c.value = value
+	return c
+}
+
+func TestCounterInc(t *testing.T) {
+	for _, tc := range []struct {
+		desc                     string
+		counter, expectedCounter uint64
+		overflow                 bool
+	}{
+		{
+			desc:            "basic 1",
+			counter:         0,
+			expectedCounter: 1,
+		},
+		{
+			desc:            "basic 2",
+			counter:         123,
+			expectedCounter: 124,
+		},
+		{
+			desc:            "almost overflow",
+			counter:         math.MaxUint64 - 1,
+			expectedCounter: math.MaxUint64,
+		},
+		{
+			desc:     "max overflow",
+			counter:  math.MaxUint64,
+			overflow: true,
+		},
+	} {
+		t.Run(tc.desc, func(t *testing.T) {
+			c := counterFromValue(tc.counter)
+			c.inc()
+			val, err := c.val()
+			if tc.overflow {
+				if err == nil {
```

Comment:
BTW, this case is never triggered. Based on the test cases, the counter will always increment. I suggest adding another field in the test vectors to specify the initial value of the counter's `hasOverflowed` field and add a test case where that is `true`. In this case, you rename the existing `overflow` to `expectedOverflow` or something like that.

Reply:
I'm not sure what you mean when you say that the case is never triggered. The last test case starts at `MaxUint64`, so wouldn't that trigger the overflow error? I renamed `overflow` to `shouldOverflow` for clarity. Please let me know if I'm missing something!

Reply:
Ah you're right. I didn't notice line 44 where you increase the counter to get it in the overflow state.

Alternatively and if you want to only test the `val()` API, then you'd start the counter with the state you wanted it at the time `val()` is called and not call `inc()`. Though, the code is fine the way it is.